### ISSUE TITLE
Lints: Update trait-related lints for improved trait-sealing analysis

### ISSUE THREAD

**Issue by obi1kenobi** [obi1kenobi/cargo-semver-checks#1105]
As of #1103, we have improved analysis of whether traits are ["sealed"](https://predr.ag/blog/is-this-trait-sealed-or-not-sealed/) or not.

The new analysis can distinguish between cases where the trait is _unconditionally sealed_ ("downstream crates will suffer compile errors if they attempt to implement it") and ones where the trait is _public API sealed_ ("downstream crates must rely on non-public API if they want to implement it, meaning they aren't covered by SemVer guarantees").

An example of a public API sealed trait is the following:
```rust
pub trait PublicAPISealed {
    #[doc(hidden)]
    type Hidden;
}
```
Here, any `impl` of this trait must provide a value for the associated type, but doing so requires using a `#[doc(hidden)]` item meaning SemVer guarantees do not apply.

Many of our trait-related lints use the older "unconditionally sealed" property, since the new "public API sealed" property didn't exist. Update each of those lints to use the new property, and add test cases to prove that public API sealed traits are handled appropriately. An example of such a test case: for the `trait_added_supertrait` lint, add a public API sealed (but not unconditionally sealed!) lint to its test crate and ensure that adding a supertrait for it does not trigger the lint as the change only affects non-public API. As another example, in the `trait_newly_sealed` lint, we want to flag lints that were previously unsealed but have become public API sealed, while not flagging traits that switch between different flavors of sealing.

Lints that need updating:
- #1107
- #1108
- #1109
- #1110
- #1111
- #1112
- #1113
- #1114
- #1115
- #1116
- #1117

**Comment by geetanshjuneja** [comment]
Hi, is this a good first issue? If yes I would like to work on this.

**Comment by obi1kenobi** [comment]
Yes, it should be straightforward enough. Make sure to read the contributing guide, except instead of adding a new lint, you'll be modifying existing ones and adding new test cases to their test crates.

Ideally, consider opening separate PRs for different sub-issues, just to make it easy to review the changes.

**Comment by obi1kenobi** [comment]
Btw @geetanshjuneja I feel like you have a pretty good handle on the process here — feel free to concurrently open as many PRs targeting the sub-issues here as you'd like. Normally it's a good idea to only have a few PRs in flight at a time, but here the sub-issues are relatively straightforward and I expect most of them will be an easy merge. My PR review time is a bit bursty anyway these days, so I'm happy to review multiple PRs back-to-back.

**Comment by geetanshjuneja** [comment]
> Btw [@geetanshjuneja](https://github.com/geetanshjuneja) I feel like you have a pretty good handle on the process here — feel free to concurrently open as many PRs targeting the sub-issues here as you'd like. 

Cool, I'll try to create PRs for the remaining sub issues as soon as I can.

**Comment by obi1kenobi** [comment]
Thank you for completely obliterating all the subtasks of this issue so quickly 🚀 You probably pulled forward the release schedule for the next cargo-semver-checks by a week right here :)
